### PR TITLE
Refine DNS relay

### PR DESCRIPTION
**Pull request by Mygod**
Beware of rookie mistakes lol.

### CODE REVIEW

**Comment by Mygod** [comment]
Wait why is `dnsrelay/mod` not barred behind `local-dns-relay` feature? I am confused.

**Comment by zonyitoo** [comment]
It should be added behind `local-dns-relay` feature.

File: src/relay/mod.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
 //! Relay server in local and server side implementations.
 
-pub mod dns;
+// pub mod dns;
```
Comment (Mygod) [inline src/relay/mod.rs:3]: This `dns` mod seems unused. Is it legacy module that is no longer used?

**Comment by Mygod** [comment]
I have more changes in mind but we can review this first. Changes in this PR:

* Respect both lists for qname matching;
* Only one DNS probe will be sent if qname was matched -- this improves privacy somewhat;
* Reverse lookup cache remembers local/remote instead of qname;
* Some optimizations.

Again beware of my rookie mistakes lol.

File: src/relay/mod.rs (hunk 1 above)
Comment (zonyitoo) [reply src/relay/mod.rs:3]: There is a binary `ssdns`, which have been removed long time ago.

Comment (Mygod) [reply src/relay/mod.rs]: Thanks! Ready to review now.

**Comment by zonyitoo** [comment]
Please merge master again.

**Comment by Mygod** [comment]
GitHub seems to think there are no conflicts?

**Comment by zonyitoo** [comment]
It said: This branch cannot be rebased due to conflicts

**Comment by Mygod** [comment]
Ah I guess rebasing is different. Let me squash everything to a single commit then.

**Comment by madeye** [comment]
This change breaks GFW list and Custom list feature in shadowsocks-android.

I'm debugging what's wrong here.

File: src/relay/dnsrelay/mod.rs [hunk 2]
```diff
@@ -123,65 +122,70 @@ async fn acl_lookup(
...
-            RData::A(ref ip) => {
-                let addr = Address::SocketAddress(SocketAddr::new(IpAddr::from(*ip), 0));
-                context.check_target_bypassed(&addr).await
-            }
-            RData::AAAA(ref ip) => {
-                let addr = Address::SocketAddress(SocketAddr::new(IpAddr::from(*ip), 0));
-                context.check_target_bypassed(&addr).await
+    let qname_in_proxy_list = context.check_qname_in_proxy_list(&addr);
+
+    let remote_response_fut = async {
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs:138]: I forgot to invoke `tokio::spawn` here. Is there any way to do this without inappropriate lifetimes?

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:138]: All the scoped libraries I found seems to swallow return values of the future, which is undesirable...

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:138]: I think previous join approach is better?

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:138]: We do not want to send out both probes all the time though. If the hostname is matched, certainly we only need to send one probe. This improves privacy slightly (use case: you want to hide even DNS traffic to some sensitive website).

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:138]: This PR looks promising, alas it is a WIP: https://github.com/tokio-rs/tokio/pull/2153/files#diff-d0c8061133c518b906f05037bd563ed7R127

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:138]: Hmm, but if we spawn here, actually we send the remote request ahead as well?

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:138]: @madeye Look at the next two lines below.

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:138]: What about this?

**Pull request merged**